pandas/core/indexes/base.py

```diff
@@ -4425,9 +4435,28 @@ def sort_values(
         """
         idx = ensure_key_mapped(self, key)
 
-        _as = idx.argsort()
-        if not ascending:
-            _as = _as[::-1]
+        # GH 35584. Sort missing values according to na_position kwarg
+        if not isinstance(self, ABCMultiIndex):
+            bad = isna(idx)
+            good = ~bad
+            _as = np.arange(len(idx), dtype=np.int64)
+            if na_position == "last":
+                _as = np.concatenate([_as[good][idx[good].argsort()], _as[bad]])
+                if not ascending:
+                    _as[: np.sum(good)] = _as[: np.sum(good)][::-1]
+            elif na_position == "first":
+                _as = np.concatenate([_as[bad], _as[good][idx[good].argsort()]])
+                if not ascending:
```

Comment:
can you share this code with the series impl (might be better / easier to create a helper to actually do this) and put it in pandas/core/sorting.py

Reply:
@jreback Found such a function in `sorting.py`: it's called `nargsort`. With a minor addition to insure that mergesort defaults to quicksort when sorting objects, it can be used for our purposes.

But I've uncovered a bit of a problem when trying to unify `Index` and `Series` behavior: when we set `ascending=False` and the object containes duplicates, in general, we expect different behavior from `Series.sort_index` and `Series.sort_values`. We expect `sort_values` to reverse the order of equal elements, but we expect `sort_index` to be stable and maintain it. This also creates weirdness in `DataFrame.sort_values`.

I've switched `Series.sort_values` to using `nargsort` and added an argument to `nargsort` to signify if we want to preserve order or not, but there are still edge cases in the test sutie because of this inconsistency.

It is probably best to settle on one convention and carefully alter all existing tests to fit it. I personally think that equal elements should maintain order when we sort in descending order, that is, the descending sort should also be stable. I don't know if we should do this in this PR though, or leave Series as it is and only add NaN-handling to `Index.sort_values` through nargsort. Then I can make another PR to handle Series there and alter the tests to the convention we agree upon.

What do you think?